templates/dynamicFee.go

```diff
@@ -0,0 +1,142 @@
+	injectionVector := []interface{}{amount, firstValid, lastValid, receiverAddr, closeRemainderAddr, lease}
+	injectedBytes, err := inject(referenceAsBytes, referenceOffsets, injectionVector)
+	if err != nil {
+		return DynamicFee{}, err
+	}
+
+	address := crypto.AddressFromProgram(injectedBytes)
+	dynamicFee := DynamicFee{
+		ContractTemplate: ContractTemplate{
+			address: address.String(),
+			program: injectedBytes,
+		},
+		receiver:       receiverAddr,
+		closeRemainder: closeRemainderAddr,
+		firstValid:     firstValid,
+		lastValid:      lastValid,
+		amount:         amount,
+		leaseBase64:    lease,
+	}
+	return dynamicFee, err
+}
+
+// GetDynamicFeeTransactions creates and signs the secondary dynamic fee transaction, updates
+// transaction fields, and signs as the fee payer; it returns both
+// transactions as bytes suitable for sendRaw.
+// Parameters:
+// txn - main transaction from payer
+// lsig - the signed logic received from the payer
+// privateKey - the private key for the account that pays the fee
+// fee - fee per byte for both transactions
+// firstValid - first protocol round on which both transactions will be valid
+// lastValid - last protocol round on which both transactions will be valid
+func GetDynamicFeeTransactions(txn types.Transaction, lsig types.LogicSig, privateKey ed25519.PrivateKey, fee, firstValid, lastValid uint64) ([]byte, error) {
+	txn.FirstValid = types.Round(firstValid)
+	txn.LastValid = types.Round(lastValid)
+	txn.Fee = types.MicroAlgos(fee)
+	eSize, err := transaction.EstimateSize(txn)
+	if err != nil {
+		return nil, err
+	}
+	txn.Fee = types.MicroAlgos(eSize * fee)
+
+	if txn.Fee < transaction.MinTxnFee {
+		txn.Fee = transaction.MinTxnFee
+	}
+
+	address := types.Address{}
+	copy(address[:], privateKey[ed25519.PublicKeySize:])
+	genesisHash := make([]byte, 32)
+	copy(genesisHash[:], txn.GenesisHash[:])
+	feePayTxn, err := transaction.MakePaymentTxn(address.String(), txn.Sender.String(), fee, uint64(txn.Fee), firstValid, lastValid, nil, "", "", genesisHash)
+	if err != nil {
+		return nil, err
+	}
+	feePayTxn.AddLease(txn.Lease, fee)
+
+	txnGroup := []types.Transaction{txn, feePayTxn}
+
+	updatedTxns, err := transaction.AssignGroupID(txnGroup, "")
+
+	_, stx1Bytes, err := crypto.SignLogicsigTransaction(lsig, updatedTxns[0])
+	if err != nil {
+		return nil, err
+	}
+	_, stx2Bytes, err := crypto.SignTransaction(privateKey, updatedTxns[1])
+	if err != nil {
+		return nil, err
+	}
+	return append(stx1Bytes, stx2Bytes...), nil
+}
+
+// SignDynamicFee returns the main transaction and signed logic needed to complete the
+// transfer. These should be sent to the fee payer, who can use
+// GetDynamicFeeTransactions() to update fields and create the auxiliary
+// transaction.
+func (c DynamicFee) SignDynamicFee(privateKey ed25519.PrivateKey, genesisHash []byte) (txn types.Transaction, lsig types.LogicSig, err error) {
```

Comment:
This one seems problematic. Do you expect the user to re-instantiate the contract? Otherwise, all the `c.element` won't exist. I'd expect this function to make a good use of `GetDynamicFeeTxns` output

Reply:
I did expect the user to re-instantiate the contract - I'll revisit the design on Confluence and look at the corresponding Python PR to see if there's a different way to do all this.

Reply:
Changing this to accept `contract []byte`, like Periodic Payment.